### PR TITLE
[Merged by Bors] - feat: Construct a tripartite graph from its triangles

### PR DESCRIPTION
**Pull request by YaelDillies**
Both the lower bound of the Ruzsa-Szemerédi problem and the proof of the corners theorem define an auxiliary tripartite graph by specifying its triangles and showing that there are no more.

This PR provides the boilerplate for this construction. It greatly simplifies both proofs.

Co-authored-by: Bhavik Mehta <bhavik.mehta8@gmail.com>

---

- [x] depends on: #12526

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:12]: 'summand' is wrong here, you mean perhaps 'component'

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:18]: 'not coming from a triangle index'

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:26]: The terminology 'Salem-Spencer set' is not really used in the literature, and is perhaps confusing - better to say just 'a set without three-term arithmetic progressions'.

File: Mathlib/Data/Finset/Card.lean
Comment (TFBloom) [inline Mathlib/Data/Finset/Card.lean:69]: All the changes in this file have already been made in #12526 and should be removed from this PR.

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Basic.lean
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Basic.lean:32]: All the changes in this file have (I think) already been made in #12526 and should be removed from this PR.

**Review by TFBloom — CHANGES_REQUESTED** [review]

File: Mathlib/Data/Finset/Card.lean
Comment (YaelDillies) [reply Mathlib/Data/Finset/Card.lean:69]: (clarified in PM)

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Basic.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Triangle/Basic.lean:32]: (clarified in PM)

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean]: Opened #12682 to fix the existing mentions of the name.

**Review by TFBloom — APPROVED** [review]
This is now all good mathematics-wise, and I recommend it be merged (pending Lean approval).

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:122]: Should the s be t here?

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:227]:
```suggestion
  (map_toTriangle_disjoint t) $ by simpa [add_assoc] using ht
```

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:208]: The prevailing style is to put the by at the end of the line (several times)

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]
Thanks, just some minor comments

**Review by jcommelin — COMMENTED** [review]
LGTM modulo Ruben's remarks

bors d+

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean:208]: As you might have noticed, ths file was hand-ported from Lean 3 😁

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**
